diff -Naurp linux/drivers/amlogic/display/osd/osd_fb.c linux-patch/drivers/amlogic/display/osd/osd_fb.c
--- linux/drivers/amlogic/display/osd/osd_fb.c	2018-06-16 05:13:17.000000000 +0300
+++ linux-patch/drivers/amlogic/display/osd/osd_fb.c	2018-06-21 13:35:07.000000000 +0300
@@ -2224,9 +2224,9 @@ static struct device_attribute osd_attrs
 			show_free_scale, store_free_scale),
 	__ATTR(scale_axis, S_IRUGO | S_IWUSR,
 			show_scale_axis, store_scale_axis),
-	__ATTR(scale_width, S_IWUSR | S_IRUGO,
+	__ATTR(scale_width, S_IRUGO | S_IRUSR,
 			show_scale_width, NULL),
-	__ATTR(scale_height, S_IWUSR | S_IRUGO,
+	__ATTR(scale_height, S_IRUGO | S_IRUSR,
 			show_scale_height, NULL),
 	__ATTR(color_key, S_IRUGO | S_IWUSR,
 			show_color_key, store_color_key),
