diff -Naurp CoreELEC-settings-238e0c2/src/resources/bin/logos.sh CoreELEC-settings-238e0c2-patch/src/resources/bin/logos.sh
--- CoreELEC-settings-238e0c2/src/resources/bin/logos.sh	2018-06-30 08:27:34.000000000 +0300
+++ CoreELEC-settings-238e0c2-patch/src/resources/bin/logos.sh	2018-06-30 08:36:59.000000000 +0300
@@ -8,6 +8,7 @@
 . /tmp/tvh-url.logos
 
 LOGOS_DIR="/storage/picons/tvh"
+RNAME_DIR="/storage/picons/rename"
 TVH_CH_COUNT=`curl -s $TVH_URL'/api/channel/grid?start=0&limit=1' | jq -r '.total'`
 TVH_CH_COUNT=`expr $TVH_CH_COUNT - 1`
 
@@ -160,6 +161,8 @@ elif [ "$1" == "missing" ] ; then
       fi
   done
 
+  rm -rf $RNAME_DIR
+  mkdir -p $RNAME_DIR
   cat $RENAME_FILE |
       while read -r RE_NAME ; do
           tmp_name=`echo "$RE_NAME" | awk -F' -\*-' '{print $1}' | sed 's/^[ \t]*//; s/[ \t]*$//'`
@@ -168,9 +171,12 @@ elif [ "$1" == "missing" ] ; then
           if [ -e "$LOGOS_DIR/$new_name.png" ] ; then
             continue
           fi
-          mv -f "$LOGOS_DIR/$old_name.png" "$LOGOS_DIR/$new_name.png"
+          cp -f "$LOGOS_DIR/$old_name.png" "$RNAME_DIR/$new_name.png"
       done
 
+  rm -rf $LOGOS_DIR
+  mv -f $RNAME_DIR $LOGOS_DIR
+
   echo "Create logos completed." > $LOG_FILE
 fi
 
