diff -Naurp CoreELEC-settings-238e0c2/src/resources/lib/modules/connman.py CoreELEC-settings-238e0c2-patch/src/resources/lib/modules/connman.py
--- CoreELEC-settings-238e0c2/src/resources/lib/modules/connman.py	2018-06-06 15:36:56.000000000 +0300
+++ CoreELEC-settings-238e0c2-patch/src/resources/lib/modules/connman.py	2018-06-22 10:39:19.000000000 +0300
@@ -633,7 +633,7 @@ class connman:
                         'wait_for_network_time': {
                             'order': 2,
                             'name': 32370,
-                            'value': '10',
+                            'value': '20',
                             'action': 'set_network_wait',
                             'type': 'num',
                             'parent': {
@@ -695,7 +695,7 @@ class connman:
             # Network Wait
 
             self.struct['advanced']['settings']['wait_for_network']['value'] = '0'
-            self.struct['advanced']['settings']['wait_for_network_time']['value'] = '10'
+            self.struct['advanced']['settings']['wait_for_network_time']['value'] = '20'
             if os.path.exists(self.WAIT_CONF_FILE):
                 wait_file = open(self.WAIT_CONF_FILE, 'r')
                 for line in wait_file:
