diff -Naurp kodi-2a11f45/xbmc/cores/VideoPlayer/DVDInputStreams/DVDInputStream.h kodi-2a11f45-patch/xbmc/cores/VideoPlayer/DVDInputStreams/DVDInputStream.h
--- kodi-2a11f45/xbmc/cores/VideoPlayer/DVDInputStreams/DVDInputStream.h	2018-12-08 12:13:56.000000000 +0200
+++ kodi-2a11f45-patch/xbmc/cores/VideoPlayer/DVDInputStreams/DVDInputStream.h	2018-12-08 12:17:18.000000000 +0200
@@ -126,7 +126,7 @@ public:
     virtual CDemuxStream* GetStream(int iStreamId) const = 0;
     virtual std::vector<CDemuxStream*> GetStreams() const = 0;
     virtual void EnableStream(int iStreamId, bool enable) {};
-    virtual bool OpenStream(int iStreamId) { return true; };
+    virtual bool OpenStream(int iStreamId) { return false; };
     virtual int GetNrOfStreams() const = 0;
     virtual void SetSpeed(int iSpeed) = 0;
     virtual bool SeekTime(double time, bool backward = false, double* startpts = NULL) = 0;
